server/src/main/java/org/opensearch/repositories/blobstore/ChecksumWritableBlobStoreFormat.java

```diff
@@ -0,0 +1,114 @@
+
+    private static final int BUFFER_SIZE = 4096;
+
+    private final String codec;
+    private final CheckedFunction<StreamInput, T, IOException> reader;
+
+    public ChecksumWritableBlobStoreFormat(String codec, CheckedFunction<StreamInput, T, IOException> reader) {
+        this.codec = codec;
+        this.reader = reader;
+    }
+
+    public BytesReference serialize(final T obj, final String blobName, final Compressor compressor) throws IOException {
+        try (BytesStreamOutput outputStream = new BytesStreamOutput()) {
+            try (
+                OutputStreamIndexOutput indexOutput = new OutputStreamIndexOutput(
+                    "ChecksumBlobStoreFormat.writeBlob(blob=\"" + blobName + "\")",
+                    blobName,
+                    outputStream,
+                    BUFFER_SIZE
+                )
+            ) {
+                CodecUtil.writeHeader(indexOutput, codec, VERSION);
+
+                try (OutputStream indexOutputOutputStream = new IndexOutputOutputStream(indexOutput) {
+                    @Override
+                    public void close() throws IOException {
+                        // this is important since some of the XContentBuilders write bytes on close.
+                        // in order to write the footer we need to prevent closing the actual index input.
+                    }
+                }; StreamOutput stream = new OutputStreamStreamOutput(compressor.threadLocalOutputStream(indexOutputOutputStream));) {
+                    // TODO The stream version should be configurable
+                    stream.setVersion(Version.CURRENT);
+                    obj.writeTo(stream);
+                }
+                CodecUtil.writeFooter(indexOutput);
+            }
+            return outputStream.bytes();
+        }
+    }
+
+    public T deserialize(String blobName, CheckedFunction<StreamInput, T, IOException> reader, BytesReference bytes) throws IOException {
+        final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+        try {
+            final IndexInput indexInput = bytes.length() > 0
+                ? new ByteBuffersIndexInput(new ByteBuffersDataInput(Arrays.asList(BytesReference.toByteBuffers(bytes))), resourceDesc)
+                : new ByteArrayIndexInput(resourceDesc, BytesRef.EMPTY_BYTES);
+            CodecUtil.checksumEntireFile(indexInput);
+            CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+            long filePointer = indexInput.getFilePointer();
+            long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+            BytesReference bytesReference = bytes.slice((int) filePointer, (int) contentSize);
+            Compressor compressor;
+            try {
+                compressor = CompressorRegistry.compressor(bytesReference);
+            } catch (NotXContentException e) {
+                compressor = CompressorRegistry.none();
```

Comment:
what happen if the compressor is not registered

Reply:
Then we would not be able to parse the content and would throw IOException

Reply:
then why are we catching the exception here, shouldnt it fail here itself

Reply:
If you look at CompressorRegistry.compressor logic here : https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/core/compress/CompressorRegistry.java#L63-L79, it assumes that the content is either compressed or it is XContent.
In our case the content may be either be compressed or it may be serialized using `writeTo` method. So we would get NotXContentException, which is fine as we will parse using `readFrom` method